### ISSUE TITLE
x509_crt_check_parent doesn't check AKID/SKID

### ISSUE THREAD

**Issue by NWilson** [ARMmbed/mbedtls#194]
We have some intermediate certificates that were re-issued with the same subject name (don't ask), so we rely on the authority key identifier to construct the chains. Unfortunately, PolarSSL doesn't check these in `x509_crt_check_parent`.

I'll make a PR for this if you'd be interested in supporting this.

Also, if I'm reading the RFC correctly, a conforming CA MUST include SKID/AKID on all generated certificates, so PolarSSL's certificate generation function should probably include it too.

RFC5280, section 4.2.1.1:
>  The authority key identifier extension provides a means of
>   identifying the public key corresponding to the private key used to
>   sign a certificate.  This extension is used where an issuer has
>   multiple signing keys (either due to multiple concurrent key pairs or
>   due to changeover).  The identification MAY be based on either the
>   key identifier (the subject key identifier in the issuer's
>   certificate) or the issuer name and serial number.
>
>   The keyIdentifier field of the authorityKeyIdentifier extension MUST
>   be included in all certificates generated by conforming CAs to
>   facilitate certification path construction.

**Comment by mpg** [comment]
Unless I'm mistaken, we're not required by the standard to check SKID/AKID, so I would tend to argue your cert chain is broken. However, I realize that you might not have enough control on the chain to avoid that, so we'll see if we can change things so that they work for you. However, I would appreciate more details about what's going on (a summary of your cert chain with simplified placeholders for names and key-ids would be great).

Regarding our behaviours when writing certs, you're right, and I'll add that on our list.

**Comment by NWilson** [comment]
You're right, you aren't required by the standard to check SKID/AKID, however, it is recommended:

> This profile RECOMMENDS support for the key identifier method by all certificate users.
>    -- (PKIX Certificate and CRL Profile, RFC5280)

I agree the cert chain is not good! We tried to be much too clever having certificates reissued, and I wish we hadn't done that. However, the SKID field is there to nasty situations work, and other SSL libraries are able to cope with it.

The cert chain that we're using here is an internal one based off our own CA cert (kept on a HSW in the server room) that we use to sign auto-updates distributed to our products in the field, which have the CA certificate built-in to them ("stapled"). Each year, we issue one new root cert and withdraw one old cert, so a product which includes 5 certs gets five years of update "coverage"; for unfortunate reasons the root certs have different names but the intermediate certs have been issued with the same subject names.

Summary: we have a bad system, which we could theoretically change, but it's easier to submit a PolarSSL PR than change it now it's deployed!

    "OU=Root cert 2014" -> "OU=Update intermediate" (SKID #1) -> leaf (monthly rotation)
    "OU=Root cert 2015" -> "OU=Update intermediate" (SKID #2) -> leaf (monthly rotation)
    
    * Updates are bundles containing a leaf cert and a signature
    * The roots and intermediates are stapled in the software
    * Finding the correct intermediate for a given leaf requires checking the
      SKID of each intermediate in the store

As far as writing certificates goes, `programs/x509/cert_write.c` already includes the SKID, so I think that's actually done, and you can take it off your list again.

Thanks very much for considering this. You can see whether you'd want the pull request -- the commit is here: https://github.com/NWilson/mbedtls/commits/akid  (As before, I know it needs tests.)

**Comment by mpg** [comment]
After a very quick first look, your patch looks good. However, I will need more time to review it carefully and write tests (or review/expand yours) before I can consider integrating it. As I said in the other issue, I'm quite pressed by other work now, so please expect a delay before I can possibly integrate this patch.

In the meantime, I would encourage you to get in touch with @pjbakker (paul dot bakker at arm dot com) regarding our contributor license agreement, so that will be one less thing to worry about later.

**Comment by NWilson** [comment]
No problem! I'm not in a hurry, thanks very much for your time so far. I imagine you're churning out stuff for ARM's requirements now to meet the goals in the mbedTLS press releases, I can see this isn't high priority.

I've emailed Paul for the legal stuff. Give me a shout whenever you want some tests written and I'll see what I can do (the PolarSSL test harness works in a strange way, it's taking me quite a while to produce each test, I'm not really an X.509 expert at all).

**Comment by NWilson** [comment]
I did make some tests for this a while back, I might rebase it to 2.0 and put together a PR sometime next month.

**Comment by mpg** [comment]
A PR against the current development branch would be appreciated. Please take care to cleanly separate the steps/features into different commits, or even two PRs, one for using key ids and one for writing them, if you choose to do both (sorry if I'm stating the obvious, we sometimes receive monolithic PRs that are hard to handle).

@pjbakker: can you please confirm if we received a CLA?

**Comment by pjbakker** [comment]
Sent out the CLA in may, we have not yet received a signed copy in return.

**Comment by NWilson** [comment]
A signed copy of the CLA (PDF scan) was sent to cla-submissions@arm.com on 28 May 2015 3:48 PM. Should I resend it, to a different address?

**Comment by pjbakker** [comment]
I correct myself. We did receive it. Thanks for providing the reference.

**Comment by davidhorstmann-arm** [comment]
Note that #6866 has done some of the work towards implementing this.
